Remaining STJ work for .NET 8 (#37806)
dotnet/docs: docs/standard/serialization/system-text-json/reflection-vs-source-generation.md

---
title: How to choose reflection or source generation in System.Text.Json
description: "Learn how to choose reflection or source generation in System.Text.Json."
ms.date: 10/30/2023
no-loc: [System.Text.Json]
zone_pivot_groups: dotnet-preview-version
---

# Reflection versus source generation in System.Text.Json

This article explains the differences between reflection and source generation as it relates to `System.Text.Json` serialization. It also provides guidance on how to choose the best approach for your scenario.

## Metadata collection

To serialize or deserialize a type, <xref:System.Text.Json.JsonSerializer> needs information about how to access the members of the type. `JsonSerializer` needs the following information:

* How to access property getters and fields for serialization.
* How to access a constructor, property setters, and fields for deserialization.
* Information about which attributes have been used to customize serialization or deserialization.
* Run-time configuration from <xref:System.Text.Json.JsonSerializerOptions>.

This information is referred to as *metadata*.

## Reflection

> By default, `JsonSerializer` collects metadata at run time by using [reflection](/dotnet/csharp/advanced-topics/reflection-and-attributes/). Whenever `JsonSerializer` has to serialize or deserialize a type for the first time, it collects and caches this metadata. The metadata collection process takes time and uses memory.

Reviewer: Let's xref the first reference to this API

```suggestion
By default, <xref:System.Text.Json.JsonSerializer> collects metadata at run time by using [reflection](/dotnet/csharp/advanced-topics/reflection-and-attributes/). Whenever `JsonSerializer` has to serialize or deserialize a type for the first time, it collects and caches this metadata. The metadata collection process takes time and uses memory.
```

## Source generation

> As an alternative, `System.Text.Json` can use the C# [source generation](../../../csharp/roslyn-sdk/source-generators-overview.md) feature to improve performance, reduce private memory usage, and facilitate [assembly trimming](../../../core/deploying/trimming/trim-self-contained.md), which reduces app size. In addition, reflection can't be used in [Native AOT applications](../../../core/deploying/native-aot/index.md), so you must use source generation for those apps.

Reviewer 2: ```suggestion
As an alternative, `System.Text.Json` can use the C# [source generation](../../../csharp/roslyn-sdk/source-generators-overview.md) feature to improve performance, reduce private memory usage, and facilitate [assembly trimming](../../../core/deploying/trimming/trim-self-contained.md), which reduces app size. In addition, certain reflection APIs can't be used in [Native AOT applications](../../../core/deploying/native-aot/index.md), so you must use source generation for those apps.
```

Source generation can be used in two modes:

* **Metadata collection mode**

>   During compilation, System.Text.Json collects the metadata needed for serialization and generates source code files. The generated source code files are automatically compiled as an integral part of the application.

Reviewer: ```suggestion
  During compilation, `System.Text.Json` collects the metadata needed for serialization and generates source code files. The generated source code files are automatically compiled as an integral part of the app.
```

>   During compilation, `System.Text.Json` collects the metadata needed for serialization and generates source code files. The generated source code files are automatically compiled as an integral part of the app.

Reviewer 2: ```suggestion
  During compilation, `System.Text.Json` collects the information needed for serialization and generates source code files that populate JSON contract metadata for the requested types.
```

* **Serialization optimization (fast track) mode**

>   <xref:System.Text.Json.JsonSerializer> features that customize the output of serialization, such as naming policies and reference preservation, carry a performance overhead. In serialization optimization mode, System.Text.Json generates optimized code that uses [`Utf8JsonWriter`](use-utf8jsonwriter.md) directly. This optimized or *fast path* code increases serialization throughput.

Reviewer 2: Consider adding a section clarifying that a fast-path _deserialization_ mode is not yet implemented:

https://github.com/dotnet/runtime/issues/55043

Source generation for `System.Text.Json` requires C# 9.0 or a later version.

## Feature comparison

Choose reflection or source-generation modes based on the following benefits that each one offers:

:::zone pivot="dotnet-8-0"

| Benefit                                              | Reflection | Source generation<br/>(Metadata collection mode) | Source generation<br/>(Serialization optimization mode) |
|------------------------------------------------------|------------|---------------------|----------------------------|

> | Simpler to code and debug.                           | ✔️        | ❌                  | ❌                        |

Reviewer: I'm really curious about this row, as I would have assumed that the _debug_ aspect would have been the same if not better for the either source generation approach over the reflection approach since you can step-debug the source generated code — which is a really nice feature.

Reviewer 2: Agreed, fast-path really has the edge when it comes to debuggability since all serialization logic is inlined in one method.

> | Supports non-public accessors.                       | ✔️        | ❌                  | ❌                        |

Reviewer 2: ```suggestion
| Supports non-public members.                       | ✔️        | ❌                  | ❌                        |
```
Starting with .NET 8, the source generator does support _some_ non-public members (e.g. internal types in the same assembly), so perhaps an emoji between ✔️ and ❌ is warranted?

> | Supports all available serialization customizations. | ✔️        | ❌                  | ❌                        |

Reviewer 2: Source generated contracts can be modified using the contract customization API.

> | Reduces start-up time.                               | ❌        | ✔️                  | ❌                        |

Reviewer 2: ```suggestion
| Reduces start-up time.                               | ❌        | ✔️                  | ✔️                        |
```

| Reduces private memory usage.                        | ❌        | ✔️                  | ✔️                        |
| Eliminates run-time reflection.                      | ❌        | ✔️                  | ✔️                        |
| Facilitates trim-safe app size reduction.            | ❌        | ✔️                  | ✔️                        |
| Increases serialization throughput.                  | ❌        | ❌                  | ✔️                        |

:::zone-end

:::zone pivot="dotnet-7-0,dotnet-6-0"

| Benefit                                              | Reflection | Source generation:<br/>Metadata collection | Source generation:<br/>Serialization optimization |
|------------------------------------------------------|------------|---------------------|----------------------------|
| Simpler to code and debug.                           | ✔️        | ❌                  | ❌                        |

> | Supports non-public accessors.                       | ✔️        | ❌                  | ❌                        |

Reviewer 2: ```suggestion
| Supports non-public members.                       | ✔️        | ❌                  | ❌                        |
```

| Supports required properties.                        | ✔️        | ❌                  | ❌                        |
| Supports init-only properties.                       | ✔️        | ❌                  | ❌                        |

> | Supports all available serialization customizations. | ✔️        | ❌                  | ❌                        |

Reviewer 2: ```suggestion
```

> | Reduces start-up time.                               | ❌        | ✔️                  | ❌                        |

Reviewer 2: ```suggestion
| Reduces start-up time.                               | ❌        | ✔️                  | ✔️                        |
```
